### DISCUSSION TITLE
Getting pool statistics

### DISCUSSION THREAD

**Discussion by NiKiZe** [mysql-net/MySqlConnector#1173] [Q&A] [answered]
Hi,
I use this hack for `MySql.Data` to get stats on the connection pool thru reflection.
Totally understanding that this is in no way supported, I still wonder if I could get some pointers on where similar values might exist in `MySqlConnector` ?

Logic psuedo explained:
* Grab ConnectionPool object using connection string, since that is the unique key
* Pull out Available "pool slots", inUse connections in pool, idle connections in pool

```C#
        public int[] GetPoolStatsViaReflection()
        {
            var asm = typeof(MySqlConnectionStringBuilder).Assembly;
            var poolManagerType = asm.GetType("MySql.Data.MySqlClient.MySqlPoolManager"); // internal sealed
            var poolType = asm.GetType("MySql.Data.MySqlClient.MySqlPool"); // internal sealed

            var pool = poolManagerType.InvokeMember("GetPool", System.Reflection.BindingFlags.InvokeMethod | System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.Public, null, null, new object[] { this.connectionStringBuilder });

            const System.Reflection.BindingFlags nonPublicInstanceField = System.Reflection.BindingFlags.GetField | System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance;
            var totalAvailable = (int)poolType.InvokeMember("available", nonPublicInstanceField, null, pool, new object[] { });
            // List<Driver>
            var inUsePool = (System.Collections.ICollection)poolType.InvokeMember("inUsePool", nonPublicInstanceField, null, pool, new object[] { });
            // Queue<Driver>
            var idlePool = (System.Collections.ICollection)poolType.InvokeMember("idlePool", nonPublicInstanceField, null, pool, new object[] { });

            return new[] {totalAvailable, inUsePool.Count, idlePool.Count};
        }
```

I guess this lives in `MySqlConnector.Core.ConnectionPool` and maybe `m_leasedSessions` is similar to inUsePool?

**Comment by bgrainger** [comment]
Related: https://github.com/mysql-net/MySqlConnector/issues/491

**Comment by bgrainger** [comment, marked as answer]
`MySqlConnector.Core.ConnectionPool.s_pools` will contain all known connection pools: https://github.com/mysql-net/MySqlConnector/blob/62ad22da8564345f8b342ec76a53023f10821c00/src/MySqlConnector/Core/ConnectionPool.cs#L600

`MySqlConnector.Core.ConnectionPool.m_sessions.Count` has the number of idle (pooled) sessions: https://github.com/mysql-net/MySqlConnector/blob/62ad22da8564345f8b342ec76a53023f10821c00/src/MySqlConnector/Core/ConnectionPool.cs#L607

`MySqlConnector.Core.ConnectionPool.m_leasedSessions.Count` has the number of active (in-use) sessions: https://github.com/mysql-net/MySqlConnector/blob/62ad22da8564345f8b342ec76a53023f10821c00/src/MySqlConnector/Core/ConnectionPool.cs#L608

`MySqlConnector.Core.ConnectionPool.ConnectionSettings.MaximumPoolSize` contains the maximum number of active connections that can be created: https://github.com/mysql-net/MySqlConnector/blob/62ad22da8564345f8b342ec76a53023f10821c00/src/MySqlConnector/Core/ConnectionPool.cs#L13 https://github.com/mysql-net/MySqlConnector/blob/62ad22da8564345f8b342ec76a53023f10821c00/src/MySqlConnector/Core/ConnectionSettings.cs#L215

I'm sure it goes without saying but **this is completely unsupported**. Moreover, this is highly concurrent code, and some of these properties are **not safe to read** when there are concurrent writers.

**Reply by NiKiZe** [reply]
Fantastic, thanks!
I have not yet seen any issue with getting counts without locks (not thinking about getting missmatching values)
Saying it again ***absolutely unsupported***, The usecase is to grab some data on exceptions. So should be rare.

Initial implementation:
```C#
            var asm = typeof(MySqlConnectionStringBuilder).Assembly;
            const string MySqlCoreName = "MySqlConnector." + nameof(MySqlConnector.Core);
            var poolType = asm.GetType(MySqlCoreName + ".ConnectionPool"); // internal sealed
            var settingType = asm.GetType(MySqlCoreName + ".ConnectionSettings"); // internal sealed

            var pool = poolType.InvokeMember("GetPool", System.Reflection.BindingFlags.InvokeMethod | System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.Public, null, null, new object[] { this.connectionStringBuilder.ConnectionString });

            const System.Reflection.BindingFlags nonPublicInstanceField = System.Reflection.BindingFlags.GetField | System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance;
            var settings = poolType.InvokeMember("ConnectionSettings", nonPublicInstanceField | System.Reflection.BindingFlags.GetProperty, null, pool, new object[] { });
            var maxPoolSize = (int)settingType.InvokeMember("MaximumPoolSize", nonPublicInstanceField | System.Reflection.BindingFlags.GetProperty, null, settings, new object[] { });
            // Dictionary<string, ServerSession> // has the number of active (in-use) sessions:
            var inUse = (System.Collections.ICollection)poolType.InvokeMember("m_leasedSessions", nonPublicInstanceField, null, pool, new object[] { });
            // LinkedList<ServerSession> // has the number of idle (pooled) sessions:
            var idle = (System.Collections.ICollection)poolType.InvokeMember("m_sessions", nonPublicInstanceField, null, pool, new object[] { });

            var inUseCount = inUse.Count;
            var idleCount = idle.Count;

            return new[] { maxPoolSize - inUseCount, inUseCount, idleCount };
```
